Core/src/ExponentialBackoff.php

```diff
@@ -49,20 +49,34 @@ class ExponentialBackoff
      */
     private $onRetryException;
 
+    /**
+     * @var callable|null
+     */
+    private $onExecutionStart;
+
     /**
      * @param int $retries [optional] Number of retries for a failed request.
      * @param callable $retryFunction [optional] returns bool for whether or not to retry
      * @param callable $onRetryException [optional] runs before the $retryFunction. Unlike the $retryFunction,
-     * this function isn't responsible to decide if a retry should happen or not,
-     * but it gives the users flexibility to consume exception messages and add custom logic.
-     * Ex: One might want to change heaers on every retry, this function can be used to achieve
-     * such a functionality.
+     *   this function isn't responsible to decide if a retry should happen or not,
+     *   but it gives the users flexibility to consume exception messages and add custom logic.
+     *   Ex: One might want to change heaers on every retry, this function can be used to achieve
+     *   such a functionality.
+     * @param callable $onExecutionStart [optional] runs before execution of the
+     *   execute function. Taken in $arguments as reference and thus gives users,
+     *   the flexibility to add custom logic before the execution of request starts.
+     *
      */
-    public function __construct($retries = null, callable $retryFunction = null, callable $onRetryException = null)
-    {
+    public function __construct(
+        $retries = null,
+        callable $retryFunction = null,
+        callable $onRetryException = null,
+        callable $onExecutionStart = null
```

Comment:
Although I like this approach as it adds flexibility. I wonder if we should just pass the initial headers in `$options` and simply use the `onRetryException` to increment the attempt count. 🤔 

CC @dwsupplee

Reply:
@yash30201 do you foresee some other use cases for `onExecutionStart`? If intent is based solely off of the utilization in the rest implementation included in this PR, I lean towards agreeing with @saranshdhingra that this might not be necessary and we could bake the logic into `onRetryException`.

Reply:
@dwsupplee, @saranshdhingra

This callback made it possible to always get the correct initial default headers(from `$request` as arguments) at the Rest.php level(i.e. client code) and use it for custom purpose. Otherwise we had only one choice of hardcoding the initial header value at the Rest.php level but this didn't seem foolproof to me as the code would break as soon as default header changes even a bit.

The default headers are added into request at core level and thus inevitably it would be required to inject some logic into core to fetch it.

Regarding the future use case🤔, I don't have a specific case in mind but this would allow developers to modify the request / options on the first request itself(with full flexibility to fetch `$request` as well as `$options`) as compared to just using `onRetryException()` to update the retries only. It's meant to be used with `onRetryException()` to access work on all the request calls.